### PR TITLE
Update naga and gfx, move the shader validation into a shader descriptor bit

### PR DESCRIPTION
**Pull request by kvark**
**Connections**
Includes https://github.com/gfx-rs/gfx/pull/3533 and a bunch of Naga things.

**Description**
Updates Naga with the new WGSL syntax, lots of fixes.
Large update of the GL backend.

**Testing**
Tested on some wgpu-rs examples.

### CODE REVIEW

File: wgpu-types/src/lib.rs [hunk 1]
```diff
@@ -451,6 +447,21 @@ bitflags::bitflags! {
     }
 }
 
+bitflags::bitflags! {
+    #[derive(Default)]
+    #[cfg_attr(feature = "trace", derive(serde::Serialize))]
+    #[cfg_attr(feature = "replay", derive(serde::Deserialize))]
+    pub struct ShaderFlags: u32 {
+        /// If enabled, `wgpu` will parse the shader with `Naga`
+        /// and validate it both internally and with regards to pipeline interface.
```
Comment (cwfitzgerald) [inline wgpu-types/src/lib.rs:456]:
```suggestion
        /// If enabled, `wgpu` will parse the shader with `Naga`
        /// and validate it both internally and with regard to the given 
        /// pipeline interface.
```

File: wgpu-types/src/lib.rs [hunk 2]
```diff
@@ -451,6 +447,21 @@ bitflags::bitflags! {
...
+        const VALIDATION = 1;
+        /// If enabled, `wgpu` will attempt to operate on `Naga` representation
+        /// of the shader module for both validation and translation into the
+        /// backend shader languages, where `gfx-hal` supports this.
```
Comment (cwfitzgerald) [inline wgpu-types/src/lib.rs:460]:
```suggestion
        /// If enabled, `wgpu` will attempt to operate on `Naga`'s internal representation
        /// of the shader module for both validation and translation into the
        /// backend shader language, on backends where `gfx-hal` supports this.
```

**Review by cwfitzgerald — APPROVED** [review]
Some documentation nit-picking, but code looks great!

File: wgpu-types/src/lib.rs (hunk 2 above)
Comment (kvark) [reply wgpu-types/src/lib.rs:460]: I'm hesitating to call it "internal" any more, since it's in Naga public API

**Comment by kvark** [comment]
bors r=cwfitzgerald

**Pull request merged**